lib/ui/compositing/scene.cc

```diff
@@ -126,15 +126,35 @@ void Scene::RasterizeToImage(uint32_t width,
 
   auto image = CanvasImage::Create();
   auto dl_image = CreateDeferredImage(
-      dart_state->IsImpellerEnabled(), layer_tree_, width, height,
-      std::move(snapshot_delegate), std::move(raster_task_runner),
-      std::move(unref_queue));
+      dart_state->IsImpellerEnabled(),
+      BuildLayerTree(width, height, pixel_ratio), std::move(snapshot_delegate),
+      std::move(raster_task_runner), std::move(unref_queue));
   image->set_image(dl_image);
   image->AssociateWithDartWrapper(raw_image_handle);
 }
 
-std::shared_ptr<flutter::LayerTree> Scene::takeLayerTree() {
-  return std::move(layer_tree_);
+std::unique_ptr<flutter::LayerTree> Scene::takeLayerTree(uint64_t width,
+                                                         uint64_t height,
+                                                         float pixel_ratio) {
+  return BuildLayerTree(width, height, pixel_ratio);
+}
+
+std::unique_ptr<LayerTree> Scene::BuildLayerTree(uint32_t width,
+                                                 uint32_t height,
+                                                 float pixel_ratio) {
+  if (!valid()) {
+    return nullptr;
+  }
+  return std::make_unique<LayerTree>(layer_tree_config_,
+                                     SkISize::Make(width, height), pixel_ratio);
+}
+
+float Scene::defaultViewPixelRatio() {
```

Comment:
This adds a new single window assumption. Can we avoid this? Otherwise, please add a TODO comment

Reply:
I've added a paragraph to explain it, and an issue to track the TODO.